### PR TITLE
Add `quantization_config` trainer argument (streamline QLoRA)

### PR DESCRIPTION
**Pull request by qgallouedec**
Adds a `quantization_config` argument to `SFTTrainer`, `DPOTrainer`, `GRPOTrainer`, `RLOOTrainer`, and `RewardTrainer`, so QLoRA no longer requires reaching into `model_init_kwargs` (or worse, manual model loading)

After:

```python
SFTTrainer(
    model="meta-llama/Llama-2-7b-hf",
    quantization_config=BitsAndBytesConfig(load_in_4bit=True),
    peft_config=LoraConfig(),
    train_dataset=dataset,
)
```

Compare with before (many ressources are written like this!):

```python
model = AutoModelForCausalLM.from_pretrained(
    "meta-llama/Llama-2-7b-hf",
    quantization_config=BitsAndBytesConfig(load_in_4bit=True),
    device_map="auto",
)
SFTTrainer(
    model=model,
    train_dataset=dataset,
    peft_config=LoraConfig(),
)
```

Before (the "right" way, but not very popular):

```python
SFTTrainer(
    model="meta-llama/Llama-2-7b-hf",
    args=SFTConfig(model_init_kwargs={"quantization_config": BitsAndBytesConfig(load_in_4bit=True)}),
    peft_config=LoraConfig(),
    train_dataset=dataset,
)
```

It sits next to `peft_config` (the other non-serializable QLoRA ingredient), flows into `from_pretrained`, and raises if also set in `args.model_init_kwargs`.

## Changes

- New `quantization_config` arg on the five trainers above (+ docstrings).
- The `trl/scripts/{sft,dpo,grpo,rloo,reward}.py` CLIs now pass it directly instead of injecting into `model_init_kwargs`.
- This drops the redundant `model_init_kwargs["device_map"] = get_kbit_device_map()` line: verified on 8×H100 that QLoRA trains identically with and without it, across transformers 4.56.2 (min supported) and 5.13; distributed runs override `device_map` to `None` anyway, and single-process runs auto-place quantized weights on the current CUDA device. See #6158 
- Updated the QLoRA example in `docs/source/peft_integration.md`.

---

> [!NOTE]
> **Medium Risk**
> Touches model loading for all major trainers and reference-model paths; behavior change is mostly API surface, but incorrect quantization or duplicate config could break QLoRA runs.
> 
> **Overview**
> Adds a **`quantization_config`** argument to **`SFTTrainer`**, **`DPOTrainer`**, **`GRPOTrainer`**, **`RLOOTrainer`**, and **`RewardTrainer`**, so QLoRA can pass **`BitsAndBytesConfig`** next to **`peft_config`** instead of pre-loading the model or stuffing **`model_init_kwargs`**.
> 
> When **`model`** is a hub id, trainers merge **`quantization_config`** into **`from_pretrained`** kwargs (and the same for reference models where applicable). Setting it in both the trainer arg and **`args.model_init_kwargs`** now **raises**; passing it with an already-instantiated model **warns** and ignores it.
> 
> **`trl/scripts/{sft,dpo,grpo,rloo,reward}.py`** and example scripts pass **`get_quantization_config(model_args)`** directly to the trainer rather than injecting into **`training_args.model_init_kwargs`**. Docs and the SFT/GRPO QLoRA notebooks are updated to use **`model=model_id`** plus **`quantization_config`** and **`model_init_kwargs`** for attention/dtype only.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 77f287d3da6a1723f13922ea0d39ecb6792d9a42. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: trl/trainer/sft_trainer.py [hunk 1]
```diff
@@ -925,6 +929,7 @@ def __init__(
         optimizers: tuple[torch.optim.Optimizer | None, torch.optim.lr_scheduler.LambdaLR | None] = (None, None),
         optimizer_cls_and_kwargs: tuple[type[torch.optim.Optimizer], dict[str, Any]] | None = None,
         preprocess_logits_for_metrics: Callable[[torch.Tensor, torch.Tensor], torch.Tensor] | None = None,
+        quantization_config: "BitsAndBytesConfig | None" = None,
```
[the remark this answers is not in the document]
Comment (qgallouedec) [reply trl/trainer/sft_trainer.py:917]: although not specifically disallowed, it would be very surprising that peft_config is used as positional arg

[the remark this answers is not in the document]
Comment (qgallouedec) [reply trl/trainer/sft_trainer.py:917]: nobody

> pass peft_config by position

**Review by sergiopaniego — COMMENTED** [review]
the example scripts/notebooks in the examples/ folder should also be reviewed and updated

**Comment by qgallouedec** [comment]
right @sergiopaniego , updated!

File: trl/trainer/dpo_trainer.py [hunk 2]
```diff
@@ -551,6 +563,11 @@ def __init__(
                     "You passed `model_init_kwargs` to the `DPOConfig`, but your model is already instantiated. "
                     "The `model_init_kwargs` will be ignored."
                 )
+            if quantization_config is not None:
+                logger.warning(
+                    "You passed `quantization_config` to the trainer, but your model is already instantiated. The "
+                    "`quantization_config` will be ignored."
+                )
```
[the remark this answers is not in the document]
Comment (qgallouedec) [reply trl/trainer/dpo_trainer.py:570]: neither value is used, there's nothing to conflict over

**Comment by qgallouedec** [comment]
I'm going to merge this one with no review, it's been open for too long
